### ISSUE TITLE
CSG.subtract yield artefact polygons sometimes

### ISSUE THREAD

**Issue by plnralex** [jscad/csg.js#93]
I am working on a prototype project currently and make heavy use of csg. Prior tests with this lib resulted in perfect results as long as the test-solids where "handcrafted" and placed around 0/0/0 coordinates. The same operations where performed on those test-solids as under runtime conditions and resulted in 100% correct results.
Now with procedurally generated solids under runtime conditions problems with subtract and cutByPlane started to appear. Interestingly even slight changes to the sizes, positions or angles of the solids resulted either in a perfect result or a problem.

The problem yields something I would describe as shadow or artefact polygons that spill over from the subtracted solid or in case of cutByPlane, seem like a mirrored reflection. These shadow-polygons are thin/one-sided, so the problem seems not to come from very-thin proximities. I guess these edge cases result from numbers that close down to the Epsilon-constants. Changing them did not change anything of the behaviour however, so I am not certain here yet.

The solids I work with are all very simple and basic cube-likes with very low poly count. Exporting them as JSON objects into a testbed resulted in the same problems there. 

I would like to contribute to this project and maybe find/solve the problem myself however I do not really understand the inner workings of this lib and could use a hint or two to get me going. Is this problem a known issue?
Secondly if this issue is yet unknown/unresolved, is there a simple way to share such problems (i.g. jsfiddle like - dump json-csg-solids onto live-test-system ) so that others can see and reproduce the problem fast?

**Comment by plnralex** [comment]
@z3dev Thank you for the informations and the quick answer.
I will try to create a good test-case for the problem described and will also dig a little deeper while doing so. As you said, the problem might point to something that affects not only subtract but could point towards a bigger picture with edge cases where numbers close down to near zero values.
I try to create the test to check for such cases.

**Comment by plnralex** [comment]
@z3dev Thank you for the informations and the quick answer.
I will try to create a good test-case for the problem described and will also dig a little deeper while doing so. As you said, the problem might point to something that affects not only subtract but could point towards a bigger picture with edge cases where numbers close down to near zero values.
I try to create the test to check for such cases.

**Comment by kaosat-dev** [comment]
thanks for the information about this issue @plnralex !
Small question : you mentioned changing the constants: how did you change them ? because normally 'very-thin proximities' is indeed 99% tied to the part of the algorithm that merges 'close' neighbour points

**Comment by plnralex** [comment]
@kaosat-dev that solves the issue, thanks for reinforcing me to double check that with much more drastic changes to the constants. Previously I did only change the numbers up and down by a small margin, to no effect. I can now reproduce the problem and its origin.

Here is how I changed the constants (src/core/constants.js)

strong change towards more precision and the problem remains but does not get more prominent
const EPS = 1e-6

strong change towards less precision and the problem disappears but everything seems to run normal
const EPS = 0.1

sweet spot where problems start to partly disappear
between const EPS = 0.00002 and const EPS = 0.00005

What I did and why it appears in my case. I use much larger numbers as units for the solids, sometimes I have solid-cubelikes that range up in the 10000_x, 4000_y, 100_z that get translated, rotated and subtracted. My initial test was with solids in the size range of 100_x_y_z and in that range everything worked perfectly. So the issue will only appears when translation and rotation meet very large solids and by that even very small deviations result in progressively larger distances/angles/areas.

**Comment by kaosat-dev** [comment]
@plnralex  those are great findings ! Thanks for providing hard numbers !
My mind is a bit hazy, but those numbers might mean that we have a combo of hitting the js precisions limits in some cases and some of it is also likely due to the implementation.
Do you think, based on your observations, that the EPS could be adjusted based on the coordinates ? 
Meh seems kindof like a hack.
Some people with way better maths skills suggested we should perhaps use an integer based geometric system to deal with these kind of issues...

**Comment by z3dev** [comment]
Or a different implementation of the BSP trees. All beyond my skills.

**Comment by plnralex** [comment]
I think kaosat-devs analysis of the problem is correct. There seems to be a combination of floating-point-precision and maybe also an implementation combination-problem, that can lead to the described issue in rare cases. This kind of reminds me of the problem with some webGl shadow-casting implementations, where you also have to tweak some numbers to find the sweet-spot for your scene because default values will often lead to somewhat strange behaviour in shadow-casting (see .shadow.bias in three.js).
A quick solution would be to change eps from constant to a setup-value and add information to the documentation that maybe the default value might need to be tweaked when working with either very tiny solids or very large solids. Only a combination of i.g. boolean operation on very large with very tiny solids might not work perfectly then.

I think that even changing the implementation of the csg-algorithm to a different approach (see https://github.com/openscad/openscad/wiki/Project%3A-Survey-of-CSG-algorithms) might not solve this issue in the end. Basically as I understand most of the current algorithms will need to make decisions for a vertex being on, behind or in front of a plane and in case of vector3 going through, move in parallel or move away from a checked plane (very simplified). Thus in the end a certain degree of degeneration is to be expected when performing rotations or translations on solids. Hence the csg.js lib having these EPS-constants to deal with this.

Up to today this lib is by far the most advanced free and simple solution that runs in the browser I know of. Looking into the near future and aiming for a better csg-lib for the browser I would say that binding the current api to an web-assembly compiled math & csg-geometric-algorithm-lib that is done in C/C++ might yield the best results in performance & quality, but might not solve the EPS-thing fully. It might be worthwhile to take a look at http://alice.loria.fr/index.php/software/3-platform/71-icesl.html and http://alice.loria.fr/software/geogram/doc/html/index.html (that already can be compiled via web assembly/emscripten).

**Comment by plnralex** [comment]
Sure thing, will do that.

**Comment by kaosat-dev** [comment]
@plnralex haha I still have 'nightmares' about the `shadow.bias` from Three.js so yes, this does seem to be a similarly annoying problem

>A quick solution would be to change eps from constant to a setup-value and add information to the documentation that maybe the default value might need to be tweaked when working with either very tiny solids or very large solids. Only a combination of i.g. boolean operation on very large with very tiny solids might not work perfectly then.

This is a very good point and something that is currently impossible to do out of the box : there are a number of 'constants' that are not settable if you are a consumer of the code (and not editing the constants manually like you did) , and some less obviously hard coded things (like the color of the subtracted item when doing boolean ops) that would benefit from making the whole lib 'root' a function that returns a 'customized' version of the lib, or perhaps all the parts of the code that rely on these constants should have some advanced options to change these values...
As @z3dev  said, we are thinking of some rather drastic changes to the lib in the future , so this needs to also be taken into account. (I really dislike that everything is tied to /conflated with the CSG implementation, good or bad in the first place).

You are also right about there not being any miracle solution (at least with a polygonal implemenation as the current one, and even some implicit modeling based one has other drawbacks).
Funny you should bring up **ICESL** I absolutely love what they are doing !
Being able to define the volumetric properties of objects with essentially shaders is INSANELY cool !  :) 

Also ditto about WASM & co, it has been on our radar, but there is so little time and so much to do :)

**Comment by profOnno** [comment]
About the EPS value. Dirk Gegorius wrote a [document](http://media.steampowered.com/apps/valve/2014/DirkGregorius_ImplementingQuickHull.pdf) about QuickHull. Page 30 (2d) and 77 (3d)  have a formula for creating an epsilon value.

**Comment by z3dev** [comment]
@profOnno Thanks. The presentation looks interesting, and loaded with information.

**Comment by z3dev** [comment]
@plnralex Sorry for the late ping. Can you provide a simple example of the issue? We are working on the boolean operations of V2 now, which are slightly different. A few more of these 'precision' issues would be very helpful.

**Comment by z3dev** [comment]
@plnralex please close this issue, and create a new issue as part of OpenJSCAD.org. His library has been extensively changed and improved as part of OpenJSCAD.org V2 project, which is now released.
